provider/balance_checker.go

```diff
@@ -113,7 +211,58 @@ loop:
 			balanceCheckResult = runner.Do(func() runner.Result {
 				return runner.NewResult(bc.doCheck(ctx))
 			})
+		case evt := <-subscriber.Events():
+			switch ev := evt.(type) {
+			case event.LeaseAddFundsMonitor:
+				bc.leases[ev.LeaseID] = nil
+				if bc.cfg.LeaseFundCheckInterval > 0 {
+					bc.runEscrowCheck(ctx, ev.LeaseID, leaseCheckCh)
+				}
+			case event.LeaseRemoveFundsMonitor:
+				tm, exists := bc.leases[ev.LeaseID]
+				if !exists {
+					break
+				}
+
+				delete(bc.leases, ev.LeaseID)
+
+				if tm != nil && !tm.Stop() {
+					<-tm.C
+				}
+			}
+		case res := <-leaseCheckCh:
+			// we may have timer fired just a heart beat ahead of lease remove event.
+			if _, exists := bc.leases[res.lid]; !exists {
+				break
+			}
+
+			if res.err != nil {
+				bc.log.Info("couldn't check lease balance. retrying in 1m", "leaseId", res.lid, "error", res.err.Error())
+				bc.leases[res.lid] = time.AfterFunc(time.Minute, func() {
+					select {
+					case <-bc.lc.Done():
+					case leaseCheckCh <- bc.doEscrowCheck(ctx, res.lid):
+					}
+				})
+				break
+			}
+
+			if res.checkAfter == 0 {
+				bc.log.Debug("lease is out of funds. closing", "leaseId", res.lid)
+
+				_ = runner.Do(func() runner.Result {
+					return runner.NewResult(nil, bc.session.Client().Tx().Broadcast(ctx, mtypes.NewMsgCloseBid(res.lid.BidID())))
```

Comment:
This is a bad idea because if we get our logic wrong we'll start closing bids too early. Just run the withdrawal transaction at this time. The network closes things if it is actually overdrawn.